### PR TITLE
nginx: upgrade pcre to pcre2

### PR DESCRIPTION
**Pull request by de11n**
Nginx builds with pcre2 by default as of version 1.21.5.

I've tested with both `nginx` and `nginxMainline`.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by emilazy** [comment]
Do we have a tracking issue for this? I learned the other day that Debian is phasing out PCRE because of security conecrns, so we should probably follow suit.

**Comment by Izorkin** [comment]
Do the lua modules work?

**Comment by emilazy** [comment]
I’ve opened https://github.com/NixOS/nixpkgs/issues/356387 to track this.

**Comment by emilazy** [comment]
@Izorkin Debian already started working on removing PCRE in 2021 and the library is EOL. The latest Debian stable release doesn’t ship it at all. I doubt anything significant is broken by this (and anything that is broken by it is probably in a dire maintenance state).

**Comment by emilazy** [comment]
@ofborg build nginx.passthru.tests

**Comment by ulrikstrid** [comment]
Should we wait for after the stable release?

**Comment by emilazy** [comment]
I think the opposite – we should backport this to 24.11 because we’re exposing a library that went EOL in 2021 and hasn’t been updated since to untrusted user input. My previous comment was wrong on the timeline: Debian considered it too risky for that purpose at the time, which is why they’ve already removed it. It’s concerning that we’ve fallen so far behind, especially for something as security‐critical as Nginx.

**Comment by ulrikstrid** [comment]
Then I think we should go for it

**Review by ulrikstrid — APPROVED** [review]

**Comment by emilazy** [comment]
## `nixpkgs-review` result                                                                                                                                       
                                                                                                                                                                 
Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).                                                                                     
                                                                                                                                                                 
Command: `nixpkgs-review pr 355989`                                                                                                                              
                                                                                                                                                                 
---                                                                                                                                                              
### `x86_64-linux`                                                                                                                                               
<details>                                                                                                                                                        
  <summary>:x: 2 packages failed to build:</summary>                                                                                                             
  <ul>                                                                                                                                                           
    <li>openresty</li>                                                                                                                                           
    <li>openresty.doc</li>                                                                                                                                       
  </ul>                                                                                                                                                          
</details>                                                                                                                                                       
<details>                                                                                                                                                        
  <summary>:white_check_mark: 17 packages built:</summary>                                                                                                       
  <ul>                                                                                                                                                           
    <li>angie</li>                                                                                                                                               
    <li>angie.doc</li>                                                                                                                                           
    <li>angieQuic</li>                                                                                                                                           
    <li>angieQuic.doc</li>                                                                                                                                       
    <li>devpi-client</li>                                                                                                                                        
    <li>devpi-client.dist</li>                                                                                                                                   
    <li>devpi-server</li>                                                                                                                                        
    <li>devpi-server.dist</li>                                                                                                                                   
    <li>nginx (nginxStable)</li>                                                                                                                                 
    <li>nginx.doc (nginxStable.doc)</li>                                                                                                                         
    <li>nginxMainline</li>                                                                                                                                       
    <li>nginxMainline.doc</li>                                                                                                                                   
    <li>nginxQuic</li>                                                                                                                                           
    <li>nginxQuic.doc</li>                                                                                                                                       
    <li>nginxShibboleth</li>                                                                                                                                     
    <li>nginxShibboleth.doc</li>                                                                                                                                 
    <li>tests.testers.lycheeLinkCheck.network</li>                                                                                                               
  </ul>                                                                                                                                                          
</details>

Looks like this breaks OpenResty:

```
./configure: error: the HTTP rewrite module requires the PCRE library.
```

We can probably backport the upstream fix.

**Comment by emilazy** [comment]
cc @thoughtpolice @lblasc @kalekseev for OpenResty

**Review by lblasc — APPROVED** [review]

**Comment by emilazy** [comment]
(Pinging because this would *break* OpenResty unless we presumably backport a patch from upstream Nginx or perhaps bump the version 😅)

**Comment by de11n** [comment]
I'll look today.

**Comment by de11n** [comment]
Upgrading openresty to `1.27.1.1` fixed the build. The NixOS tests also pass.

**Review by emilazy — APPROVED** [review]
Sounds good to me! Looks like it should be backwards‐compatible enough for the backport. I’ll let @lblasc take another look before merging.

**Comment by lblasc** [comment]
@emilazy sorry, I've reacted too fast on your comment, best way to handle openresty is to bump the version, they don't brake compatibility, should be safe for backporting.

@de11n thx for doing the bump.

**Review by lblasc — APPROVED** [review]

**Comment by SuperSandro2000** [comment]
With this change my nginx fails to start with bad system call. I confirmed this by deploying it with an override on pcre2 to pcre.

```
systemd[1]: nginx.service: Control process exited, code=exited, status=159/n/a
nginx-pre-start[4083150]: /nix/store/pjw413fdjq9fmnr718c6y3mai5v1bgng-unit-script-nginx-pre-start/bin/nginx-pre-start: line 7: 4083151 Bad system call         (core
 dumped) /nix/store/gs6s5r211gbd2banxkfxq80j4nmx6jbh-nginxQuic-1.27.2/bin/nginx -c /etc/nginx/nginx.conf -t
systemd[1]: nginx.service: Failed with result 'exit-code'.
```

dmesg contains:

```
audit: type=1326 audit(1732833069.740:14): auid=4294967295 uid=60 gid=60 ses=4294967295 subj=kernel pid=4082877 comm="nginx" exe="/nix/store/gs6s5r211gbd2banxkfxq
80j4nmx6jbh-nginxQuic-1.27.2/bin/nginx" sig=31 arch=c000003e syscall=319 compat=0 ip=0x7fc61bf4ddcb code=0x80000000
audit: type=1326 audit(1732833080.876:15): auid=4294967295 uid=60 gid=60 ses=4294967295 subj=kernel pid=4083151 comm="nginx" exe="/nix/store/gs6s5r211gbd2banxkfxq
80j4nmx6jbh-nginxQuic-1.27.2/bin/nginx" sig=31 arch=c000003e syscall=319 compat=0 ip=0x7f7a0ba68dcb code=0x80000000
```

syscall 319 maps to memfd_create.

It feels similar to https://github.com/NixOS/nixpkgs/pull/179444 but neither disabling MemoryDenyWriteExecute nor allowing memfd_create fixed the crash.

**Comment by SuperSandro2000** [comment]
I did a fix in #360008

**Pull request merged**
